tests/returntypes/gradients/test_pulse_gradient_new.py

```diff
@@ -0,0 +1,900 @@
+        res = fn(qml.execute(tapes, dev, None))
+        assert isinstance(res, tuple) and len(res) == 2
+        assert qml.math.allclose(res[0][0], np.zeros(5))
+        assert qml.math.allclose(res[1][0], np.zeros((2, 5)))
+
+    @pytest.mark.parametrize("num_samples", [1, 3])
+    @pytest.mark.parametrize("t", [2.0, 3, (0.5, 0.6)])
+    def test_constant_rx(self, num_samples, t):
+        """Test that the derivative of a pulse generated by a constant Hamiltonian
+        is computed correctly."""
+        import jax
+
+        jax.config.update("jax_enable_x64", True)
+        jnp = jax.numpy
+        params = [jnp.array(0.24)]
+        T = t if isinstance(t, tuple) else (0, t)
+        ham_single_q_const = qml.pulse.constant * qml.PauliY(0)
+        op = qml.evolve(ham_single_q_const)(params, t)
+        tape = qml.tape.QuantumScript([op], [qml.expval(qml.PauliZ(0))])
+
+        dev = qml.device("default.qubit.jax", wires=1)
+        # Effective rotation parameter
+        p = params[0] * (T[1] - T[0])
+        r = qml.execute([tape], dev, None)
+        assert qml.math.isclose(r, jnp.cos(2 * p), atol=1e-4)
+        tapes, fn = stoch_pulse_grad(tape, num_samples=num_samples)
+        assert len(tapes) == num_samples * 2
+
+        res = fn(qml.execute(tapes, dev, None))
+        assert qml.math.isclose(res, -2 * jnp.sin(2 * p) * (T[1] - T[0]))
+
+    @pytest.mark.parametrize("t", [0.02, (0.5, 0.6)])
+    def test_sin_envelope_rx_expval(self, t):
+        """Test that the derivative of a pulse with a sine wave envelope
+        is computed correctly when returning an expectation value."""
+        import jax
+
+        jnp = jax.numpy
+
+        T = t if isinstance(t, tuple) else (0, t)
+
+        dev = qml.device("default.qubit.jax", wires=1)
+        params = [jnp.array([2.3, -0.245])]
+
+        ham = self.sine * qml.PauliZ(0)
+        op = qml.evolve(ham)(params, t)
+        tape = qml.tape.QuantumScript([qml.Hadamard(0), op], [qml.expval(qml.PauliX(0))])
+
+        # Effective rotation parameter
+        x, y = params[0]
+        theta = -x / y * (jnp.cos(y * T[1]) - jnp.cos(y * T[0]))
+        theta_jac = jnp.array(
+            [
+                theta / x,
+                x / y**2 * (jnp.cos(y * T[1]) - jnp.cos(y * T[0]))
+                + x / y * (jnp.sin(y * T[1]) * T[1] - jnp.sin(y * T[0]) * T[0]),
+            ]
+        )
+        r = qml.execute([tape], dev, None)
+        assert qml.math.isclose(r, jnp.cos(2 * theta))
+
+        tapes, fn = stoch_pulse_grad(tape, num_samples=100)
+        assert len(tapes) == 200
+
+        res = fn(qml.execute(tapes, dev, None))
+        exp_grad = -2 * jnp.sin(2 * theta) * theta_jac
+        # classical Jacobian is being estimated with the Monte Carlo sampling -> coarse tolerance
+        assert qml.math.allclose(res, exp_grad, atol=0.2)
+
+    @pytest.mark.parametrize("t", [0.02, (0.5, 0.6)])
+    def test_sin_envelope_rx_probs(self, t):
```

Comment:
I think it is good to have because of the new return type system and the additional axis in the output to be handled. :thinking: Does that make sense?

Reply:
Ah sorry I thought they were the same tests just with exchanging `Z` and `X`, didnt see one was with expval and one with probs. All good 👍